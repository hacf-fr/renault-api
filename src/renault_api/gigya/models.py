"""Gigya models."""
from dataclasses import dataclass
from typing import Any
from typing import Dict
from typing import List
from typing import Optional

from . import exceptions
from renault_api.models import BaseModel

COMMON_ERRRORS: List[Dict[str, Any]] = [
    {
        "errorCode": 403042,
        "error_type": exceptions.InvalidCredentialsException,
    }
]


@dataclass
class GigyaResponse(BaseModel):
    """Gigya response."""

    errorCode: int
    errorDetails: Optional[str]

    def raise_for_error_code(self) -> None:
        """Checks the response information."""
        if self.errorCode > 0:
            for common_error in COMMON_ERRRORS:
                if self.errorCode == common_error["errorCode"]:
                    error_type = common_error["error_type"]
                    raise error_type(self.errorCode, self.errorDetails)
            raise exceptions.GigyaResponseException(self.errorCode, self.errorDetails)


@dataclass
class GigyaLoginSessionInfo(BaseModel):
    """Gigya Login sessionInfo details."""

    cookieValue: Optional[str]


@dataclass
class GigyaLoginResponse(GigyaResponse):
    """Gigya response to POST on /accounts.login."""

    sessionInfo: Optional[GigyaLoginSessionInfo]

    def get_session_cookie(self) -> str:
        """Return cookie value from session information."""
        if not self.sessionInfo:  # pragma: no cover
            raise exceptions.GigyaException("`sessionInfo` is None in Login response.")
        if not self.sessionInfo.cookieValue:  # pragma: no cover
            raise exceptions.GigyaException(
                "`sessionInfo.cookieValue` is None in Login response."
            )
        return self.sessionInfo.cookieValue


@dataclass
class GigyaGetAccountInfoData(BaseModel):
    """Gigya GetAccountInfo data details."""

    personId: Optional[str]


@dataclass
class GigyaGetAccountInfoResponse(GigyaResponse):
    """Gigya response to POST on /accounts.getAccountInfo."""

    data: Optional[GigyaGetAccountInfoData]

    def get_person_id(self) -> str:
        """Return person id."""
        if not self.data:  # pragma: no cover
            raise exceptions.GigyaException(
                "`data` is None in GetAccountInfo response."
            )
        if not self.data.personId:  # pragma: no cover
            raise exceptions.GigyaException(
                "`data.personId` is None in GetAccountInfo response."
            )
        return self.data.personId


@dataclass
class GigyaGetJWTResponse(GigyaResponse):
    """Gigya response to POST on /accounts.getJWT."""

    id_token: Optional[str]

    def get_jwt(self) -> str:
        """Return jwt token."""
        if not self.id_token:  # pragma: no cover
            raise exceptions.GigyaException("`id_token` is None in GetJWT response.")
        return self.id_token
