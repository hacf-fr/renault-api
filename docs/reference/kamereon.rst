Kamereon low-level API Reference
================================

.. contents::
    :local:
    :backlinks: none


Core
----

.. automodule:: renault_api.kamereon
   :members:

Data models
-----------

.. automodule:: renault_api.kamereon.models
   :members:
