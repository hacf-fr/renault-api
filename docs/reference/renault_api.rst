Renault high-level API Reference
================================

.. contents::
    :local:
    :backlinks: none


Session
---------------

.. automodule:: renault_api.renault_session
   :members:

Client
--------------

.. automodule:: renault_api.renault_client
   :members:

Account
---------------

.. automodule:: renault_api.renault_account
   :members:

Vehicle
---------------

.. automodule:: renault_api.renault_vehicle
   :members:
