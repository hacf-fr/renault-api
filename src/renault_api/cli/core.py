"""Core functions for the CLI."""
import logging
import os
from datetime import datetime
from locale import getdefaultlocale
from typing import Optional

import click

from renault_api.const import CONF_LOCALE, PERMANENT_KEYS
from renault_api.credential_store import CredentialStore
from renault_api.credential_store import FileCredentialStore
from renault_api.exceptions import RenaultException
from renault_api.helpers import get_api_keys
from renault_api.model.credential import Credential

CONF_ACCOUNT_ID = "accound-id"
CONF_VIN = "vin"


class CLICredentialStore:
    """Singleton of the CredentialStore for the CLI."""

    __instance: CredentialStore = None

    def __new__(cls):
        """Get singleton Credential Store."""
        if not CLICredentialStore.__instance:
            CLICredentialStore.__instance = FileCredentialStore(
                os.path.expanduser("~/.credentials/renault-api.json")
            )
        return CLICredentialStore.__instance


def set_debug(debug: bool, log: bool) -> None:
    """Renault CLI."""
    if debug or log:
        renault_log = logging.getLogger("renault_api")
        renault_log.setLevel(logging.DEBUG)

        if log:
            # create formatter and add it to the handlers
            formatter = logging.Formatter(
                "%(asctime)s - %(name)s - %(levelname)s - %(message)s"
            )

            # create file handler which logs even debug messages
            fh = logging.FileHandler(f"logs/{datetime.today():%Y-%m-%d}.log")
            fh.setLevel(logging.DEBUG)
            fh.setFormatter(formatter)

            # And enable our own debug logging
            renault_log.addHandler(fh)

        if debug:
            logging.basicConfig()

        renault_log.warning(
            "Debug output enabled. Logs may contain personally identifiable "
            "information and account credentials! Be sure to sanitise these logs "
            "before sending them to a third party or posting them online."
        )


async def set_options(
    websession, locale: Optional[str], account: Optional[str], vin: Optional[str]
) -> None:
    """Set configuration keys."""
    await set_options(locale, account, vin)
    if locale:
        await set_locale(websession, locale)
    if account:
        credential_store = CLICredentialStore()
        credential_store[CONF_ACCOUNT_ID] = Credential(account)
    if vin:
        credential_store = CLICredentialStore()
        credential_store[CONF_VIN] = Credential(vin)


async def ensure_locale(websession) -> None:
    """Prompt the user for locale."""
    credential_store = CLICredentialStore()
    if CONF_LOCALE in credential_store:
        return

    default_locale = getdefaultlocale()[0]
    while True:
        locale = click.prompt("Please select a locale", default=default_locale)
        try:
            await set_locale(websession, locale)
        except RenaultException as exc:
            click.echo(str(exc), err=True)
        else:
            return
        click.echo(f"Locale `{locale}` is unknown.", err=True)


async def set_locale(websession, locale: str) -> None:
    """Prompt the user for locale."""
    await get_api_keys(locale, aiohttp_session=websession)

    credential_store = CLICredentialStore()
    credential_store[CONF_LOCALE] = Credential(locale)


async def display_keys(websession) -> None:
    """Get the current configuration keys."""
    await ensure_locale(websession)
    credential_store = CLICredentialStore()
    for key in PERMANENT_KEYS:
        click.echo(f"Current {key}: {credential_store.get_value(key)}")
