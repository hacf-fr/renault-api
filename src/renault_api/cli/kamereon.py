"""Singletons for the CLI."""
import click
from tabulate import tabulate

from .core import CLICredentialStore
from .core import CONF_ACCOUNT_ID
from .core import CONF_VIN
from .core import ensure_locale
from renault_api.const import CONF_LOCALE
from renault_api.exceptions import RenaultException
from renault_api.helpers import get_api_keys
from renault_api.kamereon import CREDENTIAL_GIGYA_LOGIN_TOKEN
from renault_api.kamereon import Kamereon


class CLIKamereon:
    """Singleton of the CredentialStore for the CLI."""

    __instance: Kamereon = None

    @classmethod
    async def get_client(cls, websession):
        """Get singleton Kamereon client."""
        if not CLIKamereon.__instance:
            await ensure_locale(websession)

            credential_store = CLICredentialStore()
            locale = credential_store.get_value(CONF_LOCALE)
            country = locale[-2:]
            api_keys = await get_api_keys(locale, aiohttp_session=websession)
            return Kamereon(
                websession, country, api_keys, credential_store=credential_store
            )
        return CLIKamereon.__instance


async def ensure_logged_in(websession) -> None:
    """Prompt the user for credentials."""
    await ensure_locale(websession)

    credential_store = CLICredentialStore()
    if CREDENTIAL_GIGYA_LOGIN_TOKEN in credential_store:
        return

    while True:
        user = click.prompt("user")
        password = click.prompt("password", hide_input=True)
        try:
            await do_login(websession, user, password)
        except RenaultException as exc:
            click.echo(f"Login failed: {exc}.", err=True)
        else:
            return


async def do_login(websession, user, password):
    """Attempt login."""
    await ensure_locale(websession)
    kamereon = await CLIKamereon.get_client(websession)
    await kamereon.login(user, password)


async def get_account(websession) -> str:
    """Prompt the user for account."""
    credential_store = CLICredentialStore()
    if CONF_ACCOUNT_ID in credential_store:
        return credential_store.get_value(CONF_ACCOUNT_ID)

    await ensure_logged_in(websession)
    kamereon = await CLIKamereon.get_client(websession)
    response = await kamereon.get_person()
    if not response.accounts:
        raise RenaultException("No account found.")
    if len(response.accounts) == 1:
        return response.accounts[0].accountId
    elif len(response.accounts) > 1:
        menu = "Multiple accounts found:\n"
        for i, account in enumerate(response.accounts):
            menu = menu + \
                f"\t[{i+1}] {account.accountId} ({account.accountType})\n"

        while True:
            i = int(click.prompt(f"{menu}Please select"))
            try:
                account = response.accounts[i - 1].accountId
            except (KeyError, IndexError) as exc:
                click.echo(f"Invalid option: {exc}.", err=True)
            else:
                click.echo("To avoid seeing this message again, use:"
                           f"> renault-api set --account {account}")
                return account


async def get_vin(websession, account):
    """Prompt the user for vin."""
    credential_store = CLICredentialStore()
    if CONF_VIN in credential_store:
        return

    await ensure_logged_in(websession)
    kamereon = await CLIKamereon.get_client(websession)
    response = await kamereon.get_vehicles(account)
    if not response.vehicleLinks:
        raise RenaultException("No vehicle found.")
    if len(response.vehicleLinks) == 1:
        return response.vehicleLinks[0].vin
    elif len(response.vehicleLinks) > 1:
        menu = "Multiple vehicles found:\n"
        for i, vehicle in enumerate(response.vehicleLinks):
            menu = menu + f"\t[{i+1}] {vehicle.vin}\n"

        while True:
            i = int(click.prompt(f"{menu}Please select"))
            try:
                return response.vehicleLinks[i - 1].vin
            except (KeyError, IndexError) as exc:
                click.echo(f"Invalid option: {exc}.", err=True)


async def display_accounts(websession) -> None:
    """Login to Renault."""
    await ensure_logged_in(websession)
    kamereon = await CLIKamereon.get_client(websession)
    response = await kamereon.get_person()
    accounts = {
        account.accountType: account.accountId for account in response.accounts}
    click.echo(tabulate(accounts.items(), headers=[
               "Type", "ID"]))
