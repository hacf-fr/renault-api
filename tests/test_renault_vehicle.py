"""Test cases for the Renault client API keys."""
from datetime import datetime
from datetime import timezone
from typing import List

import aiohttp
import pytest
from aioresponses import aioresponses
from aioresponses.core import RequestCall  # type:ignore
from tests import fixtures
from tests.const import TEST_ACCOUNT_ID
from tests.const import TEST_COUNTRY
from tests.const import TEST_LOCALE_DETAILS
from tests.const import TEST_VIN
from tests.test_credential_store import get_logged_in_credential_store
from tests.test_renault_session import get_logged_in_session
from yarl import URL

from renault_api.kamereon.enums import ChargeMode
from renault_api.kamereon.models import ChargeSchedule
from renault_api.renault_vehicle import RenaultVehicle


@pytest.fixture
def vehicle(websession: aiohttp.ClientSession) -> RenaultVehicle:
    """Fixture for testing RenaultVehicle."""
    return RenaultVehicle(
        account_id=TEST_ACCOUNT_ID,
        vin=TEST_VIN,
        session=get_logged_in_session(websession),
    )


def test_init(websession: aiohttp.ClientSession) -> None:
    """Test RenaultVehicle initialisation."""
    assert RenaultVehicle(
        account_id=TEST_ACCOUNT_ID,
        vin=TEST_VIN,
        session=get_logged_in_session(websession),
    )

    assert RenaultVehicle(
        account_id=TEST_ACCOUNT_ID,
        vin=TEST_VIN,
        websession=websession,
        country=TEST_COUNTRY,
        locale_details=TEST_LOCALE_DETAILS,
        credential_store=get_logged_in_credential_store(),
    )


@pytest.mark.asyncio
async def test_get_battery_status(
    vehicle: RenaultVehicle, mocked_responses: aioresponses
) -> None:
    """Test get_battery_status."""
    fixtures.inject_get_battery_status(mocked_responses)
    assert await vehicle.get_battery_status()


@pytest.mark.asyncio
async def test_get_location(
    vehicle: RenaultVehicle, mocked_responses: aioresponses
) -> None:
    """Test get_location."""
    fixtures.inject_get_location(mocked_responses)
    assert await vehicle.get_location()


@pytest.mark.asyncio
async def test_get_hvac_status(
    vehicle: RenaultVehicle, mocked_responses: aioresponses
) -> None:
    """Test get_hvac_status."""
    fixtures.inject_get_hvac_status(mocked_responses)
    assert await vehicle.get_hvac_status()


@pytest.mark.asyncio
async def test_get_charge_mode(
    vehicle: RenaultVehicle, mocked_responses: aioresponses
) -> None:
    """Test get_charge_mode."""
    fixtures.inject_get_charge_mode(mocked_responses)
    assert await vehicle.get_charge_mode()


@pytest.mark.asyncio
async def test_get_cockpit(
    vehicle: RenaultVehicle, mocked_responses: aioresponses
) -> None:
    """Test get_cockpit."""
    fixtures.inject_get_cockpit(mocked_responses, "zoe")
    assert await vehicle.get_cockpit()


@pytest.mark.asyncio
async def test_get_lock_status(
    vehicle: RenaultVehicle, mocked_responses: aioresponses
) -> None:
    """Test get_lock_status."""
    fixtures.inject_get_lock_status(mocked_responses)
    assert await vehicle.get_lock_status()


@pytest.mark.asyncio
async def test_get_charging_settings(
    vehicle: RenaultVehicle, mocked_responses: aioresponses
) -> None:
    """Test get_charging_settings."""
    fixtures.inject_get_charging_settings(mocked_responses)
    assert await vehicle.get_charging_settings()


@pytest.mark.asyncio
async def test_get_notification_settings(
    vehicle: RenaultVehicle, mocked_responses: aioresponses
) -> None:
    """Test get_notification_settings."""
    fixtures.inject_get_notification_settings(mocked_responses)
    assert await vehicle.get_notification_settings()


@pytest.mark.asyncio
async def test_get_charge_history_month(
    vehicle: RenaultVehicle, mocked_responses: aioresponses
) -> None:
    """Test get_charge_history."""
    fixtures.inject_get_charge_history(mocked_responses, "202010", "202011", "month")
    assert await vehicle.get_charge_history(
        start=datetime(2020, 10, 1),
        end=datetime(2020, 11, 15),
        period="month",
    )


@pytest.mark.asyncio
async def test_get_charge_history_day(
    vehicle: RenaultVehicle, mocked_responses: aioresponses
) -> None:
    """Test get_charge_history."""
    fixtures.inject_get_charge_history(mocked_responses, "20201001", "20201115", "day")
    assert await vehicle.get_charge_history(
        start=datetime(2020, 10, 1),
        end=datetime(2020, 11, 15),
        period="day",
    )


@pytest.mark.asyncio
async def test_get_charges(
    vehicle: RenaultVehicle, mocked_responses: aioresponses
) -> None:
    """Test get_charges."""
    fixtures.inject_get_charges(mocked_responses, "20201001", "20201115")
    assert await vehicle.get_charges(
        start=datetime(2020, 10, 1),
        end=datetime(2020, 11, 15),
    )


@pytest.mark.asyncio
async def test_get_hvac_history(
    vehicle: RenaultVehicle, mocked_responses: aioresponses
) -> None:
    """Test get_hvac_history."""
    fixtures.inject_get_hvac_history(mocked_responses, "202010", "202011", "month")
    assert await vehicle.get_hvac_history(
        start=datetime(2020, 10, 1),
        end=datetime(2020, 11, 15),
        period="month",
    )


@pytest.mark.asyncio
async def test_get_hvac_sessions(
    vehicle: RenaultVehicle, mocked_responses: aioresponses
) -> None:
    """Test get_hvac_sessions."""
    fixtures.inject_get_hvac_sessions(mocked_responses, "20201001", "20201115")
    assert await vehicle.get_hvac_sessions(
        start=datetime(2020, 10, 1),
        end=datetime(2020, 11, 15),
    )


@pytest.mark.asyncio
async def test_set_ac_start(
    vehicle: RenaultVehicle, mocked_responses: aioresponses
) -> None:
    """Test set_ac_start."""
    url = fixtures.inject_set_hvac_start(mocked_responses, "start")
    assert await vehicle.set_ac_start(
        21, datetime(2020, 11, 24, 6, 30, tzinfo=timezone.utc)
    )

    expected_json = {
        "data": {
            "type": "HvacStart",
            "attributes": {
                "action": "start",
                "targetTemperature": 21,
                "startDateTime": "2020-11-24T06:30:00Z",
            },
        }
    }

    request: RequestCall = mocked_responses.requests[("POST", URL(url))][0]
    assert expected_json == request.kwargs["json"]


@pytest.mark.asyncio
async def test_set_ac_stop(
    vehicle: RenaultVehicle, mocked_responses: aioresponses
) -> None:
    """Test set_ac_stop."""
    url = fixtures.inject_set_hvac_start(mocked_responses, "cancel")
    assert await vehicle.set_ac_stop()

    expected_json = {"data": {"type": "HvacStart", "attributes": {"action": "cancel"}}}

    request: RequestCall = mocked_responses.requests[("POST", URL(url))][0]
    assert expected_json == request.kwargs["json"]


@pytest.mark.asyncio
async def test_set_charge_mode(
    vehicle: RenaultVehicle, mocked_responses: aioresponses
) -> None:
    """Test set_charge_mode."""
    url = fixtures.inject_set_charge_mode(mocked_responses, "schedule_mode")
    assert await vehicle.set_charge_mode(ChargeMode.SCHEDULE_MODE)

    expected_json = {
        "data": {"type": "ChargeMode", "attributes": {"action": "schedule_mode"}}
    }

    request: RequestCall = mocked_responses.requests[("POST", URL(url))][0]
    assert expected_json == request.kwargs["json"]


@pytest.mark.asyncio
async def test_set_charge_schedules(
    vehicle: RenaultVehicle, mocked_responses: aioresponses
) -> None:
    """Test set_charge_schedules."""
    url = fixtures.inject_set_charge_schedule(mocked_responses, "schedules")

    schedules: List[ChargeSchedule] = []
    assert await vehicle.set_charge_schedules(schedules)

    expected_json = {
        "data": {"type": "ChargeSchedule", "attributes": {"schedules": []}}
    }

    request: RequestCall = mocked_responses.requests[("POST", URL(url))][0]
    assert expected_json == request.kwargs["json"]


@pytest.mark.asyncio
async def test_set_charge_start(
    vehicle: RenaultVehicle, mocked_responses: aioresponses
) -> None:
    """Test set_charge_start."""
    url = fixtures.inject_set_charging_start(mocked_responses, "start")

    expected_json = {
        "data": {"type": "ChargingStart", "attributes": {"action": "start"}}
    }

    assert await vehicle.set_charge_start()
    request: RequestCall = mocked_responses.requests[("POST", URL(url))][0]
    assert expected_json == request.kwargs["json"]
