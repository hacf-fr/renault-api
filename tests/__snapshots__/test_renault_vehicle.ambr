# serializer version: 1
# name: test_get_endpoints[tests/fixtures/kamereon/vehicles/captur_ii.1.json]
  dict({
    'hvac-status': None,
  })
# ---
# name: test_get_endpoints[tests/fixtures/kamereon/vehicles/captur_ii.2.json]
  dict({
    'hvac-status': None,
  })
# ---
# name: test_get_endpoints[tests/fixtures/kamereon/vehicles/clio_v.1.json]
  dict({
    'hvac-status': None,
  })
# ---
# name: test_get_endpoints[tests/fixtures/kamereon/vehicles/duster.1.json]
  dict({
    'actions/charge-set-mode': EndpointDefinition(endpoint='/kca/car-adapter/v1/cars/{vin}/actions/charge-mode', mode='default'),
    'actions/charge-set-schedule': EndpointDefinition(endpoint='/kca/car-adapter/v2/cars/{vin}/actions/charge-schedule', mode='default'),
    'actions/hvac-set-schedule': EndpointDefinition(endpoint='/kca/car-adapter/v2/cars/{vin}/actions/hvac-schedule', mode='default'),
    'actions/hvac-start': EndpointDefinition(endpoint='/kca/car-adapter/v1/cars/{vin}/actions/hvac-start', mode='default'),
    'actions/hvac-stop': EndpointDefinition(endpoint='/kca/car-adapter/v1/cars/{vin}/actions/hvac-start', mode='default'),
    'alerts': EndpointDefinition(endpoint='/vehicles/{vin}/alerts', mode='default'),
    'battery-status': EndpointDefinition(endpoint='/kca/car-adapter/v2/cars/{vin}/battery-status', mode='default'),
    'charge-history': EndpointDefinition(endpoint='/kca/car-adapter/v1/cars/{vin}/charge-history', mode='default'),
    'charge-mode': EndpointDefinition(endpoint='/kca/car-adapter/v1/cars/{vin}/charge-mode', mode='default'),
    'charge-schedule': EndpointDefinition(endpoint='/kca/car-adapter/v1/cars/{vin}/charge-schedule', mode='default'),
    'charges': EndpointDefinition(endpoint='/kca/car-adapter/v1/cars/{vin}/charges', mode='default'),
    'charging-settings': EndpointDefinition(endpoint='/kca/car-adapter/v1/cars/{vin}/charging-settings', mode='default'),
    'cockpit': EndpointDefinition(endpoint='/kca/car-adapter/v1/cars/{vin}/cockpit', mode='default'),
    'hvac-history': EndpointDefinition(endpoint='/kca/car-adapter/v1/cars/{vin}/hvac-history', mode='default'),
    'hvac-sessions': EndpointDefinition(endpoint='/kca/car-adapter/v1/cars/{vin}/hvac-sessions', mode='default'),
    'hvac-settings': EndpointDefinition(endpoint='/kca/car-adapter/v1/cars/{vin}/hvac-settings', mode='default'),
    'hvac-status': EndpointDefinition(endpoint='/kca/car-adapter/v1/cars/{vin}/hvac-status', mode='default'),
    'location': EndpointDefinition(endpoint='/kca/car-adapter/v1/cars/{vin}/location', mode='default'),
    'lock-status': EndpointDefinition(endpoint='/kca/car-adapter/v1/cars/{vin}/lock-status', mode='default'),
    'notification-settings': EndpointDefinition(endpoint='/kca/car-adapter/v1/cars/{vin}/notification-settings', mode='default'),
    'pressure': EndpointDefinition(endpoint='/kca/car-adapter/v1/cars/{vin}/pressure', mode='default'),
    'res-state': EndpointDefinition(endpoint='/kca/car-adapter/v1/cars/{vin}/res-state', mode='default'),
    'soc-levels': EndpointDefinition(endpoint='/kcm/v1/vehicles/{vin}/ev/soc-levels', mode='default'),
  })
# ---
# name: test_get_endpoints[tests/fixtures/kamereon/vehicles/megane_e-tech.1.json]
  dict({
    'lock-status': None,
  })
# ---
# name: test_get_endpoints[tests/fixtures/kamereon/vehicles/megane_e-tech.2.json]
  dict({
    'lock-status': None,
  })
# ---
# name: test_get_endpoints[tests/fixtures/kamereon/vehicles/renault_5.1.json]
  dict({
    'charge-schedule': EndpointDefinition(endpoint='/kcm/v1/vehicles/{vin}/ev/settings', mode='default'),
  })
# ---
# name: test_get_endpoints[tests/fixtures/kamereon/vehicles/spring.1.json]
  dict({
    'charge-mode': None,
    'lock-status': None,
    'res-state': None,
  })
# ---
# name: test_get_endpoints[tests/fixtures/kamereon/vehicles/twingo_ze.1.json]
  dict({
    'lock-status': None,
    'res-state': None,
  })
# ---
# name: test_get_endpoints[tests/fixtures/kamereon/vehicles/zoe_40.1.json]
  dict({
    'battery-status': EndpointDefinition(endpoint='/kca/car-adapter/v2/cars/{vin}/battery-status', mode='default'),
    'charge-mode': EndpointDefinition(endpoint='/kca/car-adapter/v1/cars/{vin}/charge-mode', mode='default'),
    'charge-schedule': EndpointDefinition(endpoint='/kca/car-adapter/v1/cars/{vin}/charge-schedule', mode='default'),
    'cockpit': EndpointDefinition(endpoint='/kca/car-adapter/v1/cars/{vin}/cockpit', mode='default'),
    'hvac-status': EndpointDefinition(endpoint='/kca/car-adapter/v1/cars/{vin}/hvac-status', mode='default'),
    'location': None,
    'lock-status': None,
    'pressure': None,
    'res-state': None,
    'soc-levels': None,
  })
# ---
# name: test_get_endpoints[tests/fixtures/kamereon/vehicles/zoe_40.2.json]
  dict({
    'battery-status': EndpointDefinition(endpoint='/kca/car-adapter/v2/cars/{vin}/battery-status', mode='default'),
    'charge-mode': EndpointDefinition(endpoint='/kca/car-adapter/v1/cars/{vin}/charge-mode', mode='default'),
    'charge-schedule': EndpointDefinition(endpoint='/kca/car-adapter/v1/cars/{vin}/charge-schedule', mode='default'),
    'cockpit': EndpointDefinition(endpoint='/kca/car-adapter/v1/cars/{vin}/cockpit', mode='default'),
    'hvac-status': EndpointDefinition(endpoint='/kca/car-adapter/v1/cars/{vin}/hvac-status', mode='default'),
    'location': None,
    'lock-status': None,
    'pressure': None,
    'res-state': None,
    'soc-levels': None,
  })
# ---
# name: test_get_endpoints[tests/fixtures/kamereon/vehicles/zoe_50.1.json]
  dict({
    'lock-status': None,
    'res-state': None,
  })
# ---
# name: test_http_get
  KamereonResponse(raw_data={'data': {'type': 'Car', 'id': 'VF1AAAAA555777999', 'attributes': {'calendar': {'monday': [{'startTime': '2300', 'duration': 480, 'activationState': False}], 'tuesday': [{'startTime': '2300', 'duration': 480, 'activationState': False}], 'wednesday': [{'startTime': '2300', 'duration': 480, 'activationState': False}], 'thursday': [{'startTime': '2300', 'duration': 480, 'activationState': False}], 'friday': [{'startTime': '2300', 'duration': 480, 'activationState': False}], 'saturday': [{'startTime': '2300', 'duration': 480, 'activationState': False}], 'sunday': [{'startTime': '2300', 'duration': 480, 'activationState': False}]}}}}, errors=None)
# ---
# name: test_http_post
  KamereonResponse(raw_data={'data': {'type': 'ChargingStart', 'id': 'guid', 'attributes': {'action': 'start'}}}, errors=None)
# ---
# name: test_http_post.1
  dict({
    'data': dict({
      'attributes': dict({
        'action': 'start',
      }),
      'type': 'ChargingStart',
    }),
  })
# ---
# name: test_set_ac_start
  dict({
    'data': dict({
      'attributes': dict({
        'action': 'start',
        'startDateTime': '2020-11-24T06:30:00Z',
        'targetTemperature': 21,
      }),
      'type': 'HvacStart',
    }),
  })
# ---
# name: test_set_ac_stop
  dict({
    'data': dict({
      'attributes': dict({
        'action': 'cancel',
      }),
      'type': 'HvacStart',
    }),
  })
# ---
# name: test_set_charge_mode
  dict({
    'data': dict({
      'attributes': dict({
        'action': 'schedule_mode',
      }),
      'type': 'ChargeMode',
    }),
  })
# ---
# name: test_set_charge_schedules
  dict({
    'data': dict({
      'attributes': dict({
        'schedules': list([
        ]),
      }),
      'type': 'ChargeSchedule',
    }),
  })
# ---
# name: test_set_charge_start
  dict({
    'data': dict({
      'attributes': dict({
        'action': 'start',
      }),
      'type': 'ChargingStart',
    }),
  })
# ---
# name: test_set_hvac_schedules
  dict({
    'data': dict({
      'attributes': dict({
        'schedules': list([
        ]),
      }),
      'type': 'HvacSchedule',
    }),
  })
# ---
